lib/OpenQA/WebAPI/Plugin/Gru.pm

```diff
@@ -59,65 +90,74 @@ sub enqueue {
+
+    my $gru = $self->schema->resultset('GruTasks')->create(
         {
             taskname => $task,
-            args     => $args,
             priority => $options->{priority} // 0,
-            run_at   => $options->{run_at} // now()});
+            args     => $args,
+            run_at   => $options->{run_at} // now(),
+            jobs     => $jobs,
+        });
+
+    $self->app->minion->enqueue(
+        $task => $args => {priority => $options->{priority} // 0, delay => $delay, notes => {gru_id => $gru->id}});
 }
 
 package OpenQA::WebAPI::Plugin::Gru::Command::gru;
 use Mojo::Base 'Mojolicious::Command';
+use Mojo::Pg;
+use Data::Dumper;
 
 has usage       => "usage: $0 gru [-o]\n";
 has description => 'Run a gru to process jobs - give -o to exit _o_nce everything is done';
 
+sub delete_gru {
+    my ($self, $id) = @_;
+    my $gru = $self->app->db->resultset('GruTasks')->find($id);
+    $gru->delete() if $gru;
+}
+
 sub cmd_list {
     my ($self) = @_;
-
-    my $tasks = $self->app->schema->resultset('GruTasks');
-    while (my $task = $tasks->next) {
-        use Data::Dumper;
-        print $task->taskname . " " . Dumper($task->args) . "\n";
+    my $tasks = $self->app->minion->backend->list_jobs();
+    foreach my $j (@{$tasks->{jobs}}) {
+        print $j->{task} . " " . Dumper($j->{args}) . " result: " . $j->{result} . "\n";
     }
 }
 
-sub run_first {
-    my ($self) = @_;
-
-    my $dtf   = $self->app->schema->storage->datetime_parser;
-    my $where = {run_at => {'<=', $dtf->format_datetime(DBIx::Class::Timestamps::now())}};
-    my $first = $self->app->schema->resultset('GruTasks')
-      ->search($where, {order_by => [{-desc => 'priority'}, 'id'], rows => 1})->first;
-
-    if ($first) {
-        $self->app->log->debug(sprintf("Running Gru task %d(%s)", $first->id, $first->taskname));
-        my $subref = $self->app->gru->{tasks}->{$first->taskname};
-        if ($subref) {
-            eval { &$subref($self->app, $first->args) };
-            if ($@) {
-                print $@ . "\n";
-                return;
-            }
-            $first->delete;
-            return 1;
-        }
+sub execute_job {
+    my ($self, $job) = @_;
+    unless (my $err = $job->_run) {
+        $job->finish;
+        $self->delete_gru($job->info->{notes}{gru_id}) if exists $job->info->{notes}{gru_id};
     }
 }
 
 sub cmd_run {
     my $self = shift;
     my $opt = $_[0] || '';
-    while (1) {
-        if (!$self->run_first) {
-            if ($opt eq '-o') {
-                return;
-            }
-            sleep(5);
+
+    my $worker = $self->app->minion->repair->worker->register;
+
+    if ($opt eq '-o') {
+        while (my $job = $worker->register->dequeue(0)) {
+            $self->execute_job($job);
         }
+        return $worker->unregister;
+    }
+
+    while (1) {
+        next unless my $job = $worker->register->dequeue(5);
+        $self->execute_job($job);
+        sleep 5;
     }
+    $worker->unregister;
```

Comment:
I wonder if this whole block could be replaced with a call to [Minion::Worker::run](http://mojolicious.org/perldoc/Minion/Worker#run).

Reply:
Jobs are performed in foreground, so that would spawn another process